### PR TITLE
Ruby: Recognise raw Erb output as XSS sink

### PR DESCRIPTION
**Pull request by hmac**
In Rails at least, `<%== x %>` is equivalent to `<%= raw x %>` or `<%= x.to_s.html_safe %>`.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ast/Erb.qll [hunk 1]
```diff
@@ -252,14 +252,32 @@ class ErbGraphqlDirective extends ErbDirective {
...
 
   override ErbCode getToken() { toGenerated(result) = g.getChild() }
 
+  /**
+   * Holds if this is a raw Erb output directive.
+   * ```erb
+   * <%== foo %>
+   * ```
+   */
+  predicate isRaw() { this = TRawOutputDirective(g) }
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Erb.qll:265]: Have you tried the following? This way you don't need any of the changes related to adding `TRawOutputDirective`

```suggestion
  predicate isRaw() {
    exists(Erb::Token t | t.getParentIndex() = 0 and t.getParent() = g and t.getValue() = "<%==")
  }
```

Comment (hmac) [reply ruby/ql/lib/codeql/ruby/ast/Erb.qll]: That's a nice simplification!

File: ruby/ql/lib/change-notes/2024-02-12-raw-erb-output.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Raw output ERb tags of the form `<%== ... %>` are now recognised as cross-site scripting sinks.
```
Comment (aibaars) [inline ruby/ql/lib/change-notes/2024-02-12-raw-erb-output.md:4]: `ERb` looks a little strange, should it be `Erb` or `ERB` ?

Comment (hmac) [reply ruby/ql/lib/change-notes/2024-02-12-raw-erb-output.md:4]: Hmm yes, I thought `ERb` was the correct casing (for "Embedded Ruby") but after some research it looks like I just invented that. Everyone else uses ERB.

**Review by aibaars — APPROVED** [review]
Looks good to me. Let's run a final DCA test.

File: ruby/ql/lib/codeql/ruby/ast/Erb.qll [hunk 3]
```diff
@@ -256,10 +256,30 @@ class ErbOutputDirective extends ErbDirective {
...
+  }
+
   final override string toString() {
-    result = "<%=" + this.getToken().toString() + "%>"
+    this.isRaw() and
+    (
+      result = "<%==" + this.getToken().toString() + "%>"
+      or
+      not exists(this.getToken()) and result = "<%==%>"
+    )
```
[the remark this answers is not in the document]
Comment (hmac) [reply ruby/ql/lib/codeql/ruby/ast/Erb.qll:275]: For reasons I don't understand, this leads to non-monotonic recursion errors.

[the remark this answers is not in the document]
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ast/Erb.qll:275]: Right. The following should work, but I think I rather prefer the current implementation
```ql
  language[monotonicAggregates]
  final override string toString() {
    result = "<%==" + concat(ErbComment tok | tok = this.getToken() | tok.toString()) + "%>"
  }
```

**Pull request merged**